File: 546. Remove Boxes.java
public class Solution {
    public int removeBoxes(int[] boxes) {
        int n = boxes.length;
        Integer[][][] dp = new Integer[n][n][n];

        return dfs(boxes, 0, n - 1, 0, dp);
    }

    private int dfs(int[] boxes, int l, int r, int k, Integer[][][] dp) {
        if (l > r) return 0;
        if (dp[l][r][k] != null) return dp[l][r][k];

        while (l < r && boxes[l] == boxes[l + 1]) {
            l++;
            k++;
        }

        int res = (k + 1) * (k + 1) + dfs(boxes, l + 1, r, 0, dp);

        for (int i = l + 1; i <= r; i++) {
            if (boxes[i] == boxes[l]) {
                res = Math.max(res, dfs(boxes, l + 1, i - 1, 0, dp) + dfs(boxes, i, r, k + 1, dp));
            }
        }

        dp[l][r][k] = res;
        return res;
    }
}

File: 546. Remove Boxes.cpp
class Solution {
public:
    int removeBoxes(vector<int>& boxes) {
        int n = boxes.size();
        int dp[100][100][100] = {0};
        
        for (int i = 0; i < n; i++) {
            for (int k = 0; k <= i; k++) {
                dp[i][i][k] = (k + 1) * (k + 1);
            }
        }
        
        for (int len = 2; len <= n; len++) {
            for (int l = 0; l + len <= n; l++) {
                int r = l + len - 1;
                for (int k = 0; k <= l; k++) {
                    // Option 1: Remove boxes[l] with its k same-colored friends
                    int res = (k + 1) * (k + 1) + dp[l + 1][r][0];
                    
                    // Option 2: Merge with a later same-colored box
                    for (int m = l + 1; m <= r; m++) {
                        if (boxes[m] == boxes[l]) {
                            res = max(res, dp[l + 1][m - 1][0] + dp[m][r][k + 1]);
                        }
                    }
                    dp[l][r][k] = res;
                }
            }
        }
        
        return dp[0][n - 1][0];
    }
};
